modules/renter/contractor/journal.go

```diff
@@ -0,0 +1,492 @@
+package contractor
+
+// The contractor achieves efficient persistence using a JSON transaction
+// journal. It enables efficient ACID transactions on JSON objects.
+//
+// Each journal represents a single JSON object. The object is serialized as
+// an "initial object" followed by a series of update sets, one per line. Each
+// update specifies a field and a modification. See the journalUpdate type for
+// a full specification.
+//
+// During operation, the object is first loaded by reading the file and
+// applying each update to the initial object. It is subsequently modified by
+// appending update sets to the file, one per line. At any time, a
+// "checkpoint" may be created, which clears the journal and starts over with
+// a new initial object. This allows for compaction of the journal file.
+//
+// In the event of power failure or other serious disruption, the most recent
+// update set may be only partially written. Partially written update sets are
+// simply ignored when reading the journal. Individual updates may also be
+// ignored if they are malformed, though other updates in the set may be
+// applied. See the journalUpdate docstring for an explanation of malformed updates.
+
+// TODO:
+// - handle case sensitivity
+// - handle null vs. empty array
+//    - if index is 0, just do a wholesale replace with [val]
+// - handle inserting new keys into object
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+)
+
+// A journal is a log of updates to a JSON object.
+type journal struct {
+	f        *os.File
+	filename string
+}
+
+// update applies the updates atomically to j. It syncs the underlying file
+// before returning.
+func (j *journal) update(us []journalUpdate) error {
+	buf := make([]byte, 0, 1024) // reasonable guess; avoids GC if we're lucky
+	buf = append(buf, '[')
+	for i, u := range us {
+		if i > 0 {
+			buf = append(buf, ',')
+		}
+		buf = append(buf, `{"p":"`...)
+		buf = append(buf, u.Path...)
+		buf = append(buf, `","v":`...)
+		buf = append(buf, *u.Value...)
+		buf = append(buf, '}')
+	}
+	buf = append(buf, ']', '\n')
+	if _, err := j.f.Write(buf); err != nil {
+		return err
+	}
+	return j.f.Sync()
+}
+
+// Checkpoint refreshes the journal with a new initial object. It syncs the
+// underlying file before returning.
+func (j *journal) checkpoint(obj interface{}) error {
+	// write to a new temp file
+	//
+	// TODO: a separate file may not be necessary. We could use an update with
+	// path "" instead, and then overwrite the beginning of the file and
+	// truncate. If the overwrite fails, we still have the full rewrite update
+	// left at the end. Just need to be careful not to overflow into the
+	// update if the new object is large.
```

Comment:
Too premature, it doubles the amount of data you write to disk. Also I think risky or even unsafe because what happens if there is corruption at the head of the file? You have to be able to read through it to find the non-corrupt part. And then I think you would not want to run the truncate operation until after syncing the write at the head, but the write at the head is quite potentially going to obliterate data anyway.

It's probably too clever, and best to wait to try to scrap the move operations until the logis properly a fully separate file.